common/solver/batch_bicgstab_kernels.hpp.inc

```diff
@@ -0,0 +1,847 @@
+                gko::batch::to_const(s_shared_entry),
+                res_norms_temp_shared_entry);  // an estimate of residual norms
+            __syncthreads();
+            copy(gko::batch::to_const(res_norms_temp_shared_entry),
+                 res_norms_shared_entry,
+                 converged);  // update res norms only for those RHSs which have
+                              // not yet converged.
+
+            __syncthreads();
+
+
+            const uint32 converged_prev = converged;
+
+            all_converged =
+                stop.check_converged(iter, res_norms_shared_entry.values,
+                                     {NULL, 0, 0, 0}, converged);
+
+            // update x for the sytems (rhs) which converge at this point...  x
+            // = x + alpha*p_hat
+
+            // note bits could change from 0 to 1, not the other way round, so
+            // we can use xor to get info about recent convergence...
+            const uint32 converged_recent = converged_prev ^ converged;
+
+            update_x_middle(
+                x_shared_entry, gko::batch::to_const(alpha_shared_entry),
+                gko::batch::to_const(p_hat_shared_entry), converged_recent);
+
+            __syncthreads();
+
+            logger.log_iteration(ibatch, iter, res_norms_shared_entry.values,
+                                 converged);
+
+            if (all_converged) {
+                break;
+            }
+
+
+            // s_hat = precond * s
+            prec_shared.apply(gko::batch::to_const(s_shared_entry),
+                              s_hat_shared_entry);
+
+            __syncthreads();
+
+
+            // t = A * s_hat
+            spmv_kernel(A_global_entry,
+                        gko::batch::to_const(s_hat_shared_entry),
+                        t_shared_entry);
+            __syncthreads();
+
+            // omega_new = <t,s> / <t,t>
+            compute_omega_new(gko::batch::to_const(t_shared_entry),
+                              gko::batch::to_const(s_shared_entry),
+                              omega_new_shared_entry, converged);
+
+            __syncthreads();
+
+            // x = x + alpha*p_hat + omega_new*s_hat
+            update_x(x_shared_entry, gko::batch::to_const(p_hat_shared_entry),
+                     gko::batch::to_const(s_hat_shared_entry),
+                     gko::batch::to_const(alpha_shared_entry),
+                     gko::batch::to_const(omega_new_shared_entry), converged);
+
+
+            // r = s - omega_new*t
+            update_r(gko::batch::to_const(s_shared_entry),
+                     gko::batch::to_const(omega_new_shared_entry),
+                     gko::batch::to_const(t_shared_entry), r_shared_entry,
+                     converged);
+
+            __syncthreads();
+
+            compute_norm2<ValueType>(
+                gko::batch::to_const(r_shared_entry),
+                res_norms_temp_shared_entry);  // residual norms
+            __syncthreads();
```

Comment:
This is only needed for the logging, right ? Can we make this optional ? This would reduce the overhead a bit, as you have a norm computation in every iteration, which seems unnecessary.

Reply:
Okay, I think we can leave it as it now for now and adapt this later, if necessary.